File: static/assets/js/components/ai-config.js
/**
 * AI配置管理组件
 * 管理AI功能的开关、模式选择和缓存管理
 */

const { createApp } = Vue;

const AIConfigApp = createApp({
    data() {
        return {
            loading: false,
            saving: false,
            clearing: false,
            training: false,
            
            // AI状态数据
            aiStatus: {
                ai_enabled: false,
                startup_mode: '',
                modules: {},
                dependencies: {},
                cache_info: {},
                lightweight_available: true,
                recommendations: []
            },
            
            // 全局配置
            globalConfig: {
                ai_mode: 'auto'
            },
            
            // 模块列表
            moduleList: [],
            
            // 配置建议
            recommendations: [],
            
            // UI状态
            statusMessage: '',
            statusType: '',
            loadingMessage: '加载中...'
        }
    },
    
    mounted() {
        this.initPage();
    },
    
    methods: {
        /**
         * 初始化页面
         */
        async initPage() {
            await this.refreshStatus();
        },
        
        /**
         * 刷新AI状态
         */
        async refreshStatus() {
            this.loading = true;
            try {
                const response = await axios.get(window.API.aiConfig.status);
                
                if (response.data.success) {
                    this.aiStatus = response.data.data;
                    this.updateModuleList();
                    this.recommendations = this.aiStatus.recommendations || [];
                    
                    // 根据当前状态推断全局模式
                    this.inferGlobalMode();
                    
                    this.showStatusMessage('状态刷新成功', 'success');
                } else {
                    throw new Error(response.data.message || '获取状态失败');
                }
            } catch (error) {
                console.error('刷新状态失败:', error);
                this.showStatusMessage('刷新状态失败: ' + (error.response?.data?.detail || error.message), 'error');
            } finally {
                this.loading = false;
            }
        },
        
        /**
         * 更新模块列表
         */
        updateModuleList() {
            this.moduleList = Object.entries(this.aiStatus.modules || {}).map(([name, config]) => ({
                name,
                description: config.description,
                enabled: config.enabled,
                mode: config.configured_mode,
                actual_mode: config.actual_mode,
                is_working: config.is_working,
                fallback_to_lightweight: config.fallback_to_lightweight
            }));
        },
        
        /**
         * 推断全局模式
         */
        inferGlobalMode() {
            const modules = Object.values(this.aiStatus.modules || {});
            
            if (modules.every(m => !m.enabled)) {
                this.globalConfig.ai_mode = 'disabled';
            } else {
                // 取最常见的模式
                const modes = modules.filter(m => m.enabled).map(m => m.configured_mode);
                const modeCount = {};
                modes.forEach(mode => {
                    modeCount[mode] = (modeCount[mode] || 0) + 1;
                });
                
                const mostCommonMode = Object.keys(modeCount).reduce((a, b) => 
                    modeCount[a] > modeCount[b] ? a : b, 'auto'
                );
                
                this.globalConfig.ai_mode = mostCommonMode;
            }
        },
        
        /**
         * 更新全局配置
         */
        async updateGlobalConfig() {
            this.saving = true;
            try {
                const response = await axios.post(window.API.aiConfig.globalConfig, this.globalConfig);
                
                if (response.data.success) {
                    this.showStatusMessage('全局配置更新成功', 'success');
                    await this.refreshStatus();
                } else {
                    throw new Error(response.data.message || '更新配置失败');
                }
            } catch (error) {
                console.error('更新全局配置失败:', error);
                this.showStatusMessage('更新全局配置失败: ' + (error.response?.data?.detail || error.message), 'error');
            } finally {
                this.saving = false;
            }
        },
        
        /**
         * 更新模块配置
         */
        async updateModuleConfig(moduleName, enabled, mode) {
            try {
                const config = {
                    module_name: moduleName,
                    enabled: enabled,
                    mode: mode
                };
                
                const response = await axios.post(window.API.aiConfig.moduleConfig, config);
                
                if (response.data.success) {
                    this.showStatusMessage(`模块 ${moduleName} 配置更新成功`, 'success');
                    // 延迟刷新状态，让配置生效
                    setTimeout(() => {
                        this.refreshStatus();
                    }, 500);
                } else {
                    throw new Error(response.data.message || '更新模块配置失败');
                }
            } catch (error) {
                console.error('更新模块配置失败:', error);
                this.showStatusMessage('更新模块配置失败: ' + (error.response?.data?.detail || error.message), 'error');
                // 回滚配置
                await this.refreshStatus();
            }
        },
        
        /**
         * 清理模型缓存
         */
        async clearModelCache() {
            try {
                if (!confirm('清理缓存将删除所有已下载的模型文件，下次使用时需要重新下载。确定要继续吗？')) {
                    return;
                }
                
                this.clearing = true;
                const response = await axios.post(window.API.aiConfig.cacheClear);
                
                if (response.data.success) {
                    const freedMB = response.data.data.freed_mb;
                    this.showStatusMessage(`缓存清理成功，释放空间 ${freedMB} MB`, 'success');
                    await this.refreshStatus();
                } else {
                    throw new Error(response.data.message || '清理缓存失败');
                }
            } catch (error) {
                console.error('清理模型缓存失败:', error);
                this.showStatusMessage('清理模型缓存失败: ' + (error.response?.data?.detail || error.message), 'error');
            } finally {
                this.clearing = false;
            }
        },
        
        /**
         * 训练轻量级模型
         */
        async trainLightweight() {
            try {
                if (!confirm('训练轻量级模型需要足够的训练数据。请确认已收集足够的广告和正常内容样本。')) {
                    return;
                }
                
                this.training = true;
                const response = await axios.post(window.API.aiConfig.lightweightTrain);
                
                if (response.data.success) {
                    this.showStatusMessage('轻量级模型训练成功', 'success');
                    await this.refreshStatus();
                } else {
                    this.showStatusMessage(response.data.message || '训练失败', 'warning');
                }
            } catch (error) {
                console.error('训练轻量级模型失败:', error);
                this.showStatusMessage('训练轻量级模型失败: ' + (error.response?.data?.detail || error.message), 'error');
            } finally {
                this.training = false;
            }
        },
        
        /**
         * 应用推荐配置
         */
        async applyRecommendation(recommendation) {
            try {
                const action = recommendation.action;
                
                if (action.type === 'install') {
                    // 显示安装命令
                    alert(`请在服务器上执行以下命令：\n\n${action.command}`);
                } else if (action.type === 'config') {
                    // 应用配置建议
                    if (action.module) {
                        // 更新单个模块
                        const module = this.moduleList.find(m => m.name === action.module);
                        if (module) {
                            module.mode = action.suggested_mode;
                            await this.updateModuleConfig(action.module, module.enabled, action.suggested_mode);
                        }
                    } else if (action.suggested_mode) {
                        // 更新全局模式
                        this.globalConfig.ai_mode = action.suggested_mode;
                        await this.updateGlobalConfig();
                    }
                }
            } catch (error) {
                console.error('应用推荐配置失败:', error);
                this.showStatusMessage('应用推荐配置失败: ' + error.message, 'error');
            }
        },
        
        /**
         * 获取模式描述
         */
        getModeDescription(mode) {
            const descriptions = {
                'auto': '自动选择最适合的模式，优先使用轻量级高效算法',
                'lightweight': '使用优化算法实现高效文本处理，无外部依赖，启动快速',
                'deep': '已弃用 - 系统已优化为轻量级模式',
                'disabled': '完全禁用功能，仅使用基础规则匹配'
            };
            return descriptions[mode] || '未知模式';
        },
        
        /**
         * 获取实际模式的标签类型
         */
        getActualModeType(mode) {
            const types = {
                'deep': 'success',
                'lightweight': 'warning',
                'rule_based': 'info',
                'disabled': 'danger'
            };
            return types[mode] || 'info';
        },
        
        /**
         * 获取实际模式的Badge类
         */
        getActualModeBadgeClass(mode) {
            const classes = {
                'deep': 'badge-success',
                'lightweight': 'badge-warning',
                'rule_based': 'badge-info',
                'disabled': 'badge-danger'
            };
            return classes[mode] || 'badge-info';
        },
        
        /**
         * 获取警告类
         */
        getAlertClass(type) {
            const classes = {
                'success': 'alert-success',
                'info': 'alert-info',
                'warning': 'alert-warning',
                'error': 'alert-danger'
            };
            return classes[type] || 'alert-info';
        },
        
        /**
         * 获取警告图标
         */
        getAlertIcon(type) {
            const icons = {
                'success': '✅',
                'info': 'ℹ️',
                'warning': '⚠️',
                'error': '❌'
            };
            return icons[type] || 'ℹ️';
        },
        
        /**
         * 显示状态消息
         */
        showStatusMessage(message, type = 'info') {
            this.statusMessage = message;
            this.statusType = type;
            
            // 3秒后自动清除消息
            setTimeout(() => {
                this.statusMessage = '';
                this.statusType = '';
            }, 3000);
        },
        
        /**
         * 返回主页
         */
        goBack() {
            window.location.href = '/static/index.html';
        }
    }
});

// 挂载应用
AIConfigApp.mount('#app');

File: app/api/ai_config.py
"""
AI配置管理API
提供Web界面管理AI功能开关和模式选择
"""
import logging
from typing import Dict, Any, List
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel

from app.core.ai_config import get_ai_config
from app.services.lightweight_similarity import get_lightweight_filter
# 临时移除权限检查，AI配置管理暂不需要特殊权限
# from app.core.auth import check_permission

logger = logging.getLogger(__name__)

from app.core.routes import ROUTES

router = APIRouter(tags=["AI配置"])

class AIModuleConfig(BaseModel):
    """AI模块配置"""
    enabled: bool
    mode: str  # auto, lightweight, deep, rule_based, disabled
    description: str

class AIConfigUpdate(BaseModel):
    """AI配置更新请求"""
    module_name: str
    enabled: bool
    mode: str

class AIGlobalConfig(BaseModel):
    """全局AI配置"""
    ai_mode: str  # auto, lightweight, deep, disabled

@router.get(ROUTES.ai.status)
async def get_ai_status():
    """获取AI功能状态"""
    try:
        ai_config = get_ai_config()
        model_cache = get_model_cache_manager()
        lightweight_filter = get_lightweight_filter()
        
        # 获取模块状态
        modules_status = {}
        for module_name, config in ai_config.ai_modules.items():
            actual_mode = ai_config.get_module_mode(module_name)
            modules_status[module_name] = {
                'description': config['description'],
                'enabled': config['enabled'],
                'configured_mode': config['mode'],
                'actual_mode': actual_mode,
                'fallback_to_lightweight': config.get('fallback_to_lightweight', False),
                'is_working': actual_mode != 'disabled'
            }
        
        # 获取依赖状态
        dependencies = {
            'sentence_transformers': False,  # 已移除
            'scikit_learn': True,  # 轻量级模式依赖
            'jieba': True  # 中文分词依赖
        }
        
        # 获取缓存信息
        cache_info = model_cache.get_cache_info()
        
        return {
            'success': True,
            'data': {
                'ai_enabled': ai_config.is_ai_enabled(),
                'startup_mode': ai_config._startup_mode,
                'modules': modules_status,
                'dependencies': dependencies,
                'cache_info': cache_info,
                'lightweight_available': True,  # 轻量级模式总是可用
                'recommendations': _get_recommendations(ai_config, dependencies)
            }
        }
        
    except Exception as e:
        logger.error(f"获取AI状态失败: {e}")
        raise HTTPException(status_code=500, detail=f"获取AI状态失败: {str(e)}")

@router.post(ROUTES.ai.config)
async def update_global_config(config: AIGlobalConfig):
    """更新全局AI配置"""
    try:
        ai_config = get_ai_config()
        
        # 验证模式
        valid_modes = ['auto', 'lightweight', 'deep', 'disabled']
        if config.ai_mode not in valid_modes:
            raise HTTPException(status_code=400, detail=f"无效的AI模式: {config.ai_mode}")
        
        # 更新所有模块配置
        for module_name in ai_config.ai_modules:
            if config.ai_mode == 'disabled':
                ai_config.ai_modules[module_name]['enabled'] = False
            else:
                ai_config.ai_modules[module_name]['enabled'] = True
                ai_config.ai_modules[module_name]['mode'] = config.ai_mode
        
        # 清除缓存
        ai_config._cache.clear()
        
        logger.info(f"全局AI配置已更新为: {config.ai_mode}")
        
        return {
            'success': True,
            'message': f'全局AI配置已更新为: {config.ai_mode}',
            'data': {
                'ai_mode': config.ai_mode,
                'affected_modules': list(ai_config.ai_modules.keys())
            }
        }
        
    except Exception as e:
        logger.error(f"更新全局AI配置失败: {e}")
        raise HTTPException(status_code=500, detail=f"更新全局AI配置失败: {str(e)}")

@router.post(ROUTES.ai.module_config)
async def update_module_config(config: AIConfigUpdate):
    """更新单个模块配置"""
    try:
        ai_config = get_ai_config()
        
        # 验证模块名
        if config.module_name not in ai_config.ai_modules:
            raise HTTPException(status_code=400, detail=f"未知的AI模块: {config.module_name}")
        
        # 验证模式
        valid_modes = ['auto', 'lightweight', 'deep', 'rule_based', 'disabled']
        if config.mode not in valid_modes:
            raise HTTPException(status_code=400, detail=f"无效的模式: {config.mode}")
        
        # 更新配置
        ai_config.ai_modules[config.module_name]['enabled'] = config.enabled
        ai_config.ai_modules[config.module_name]['mode'] = config.mode
        
        # 清除该模块的缓存
        cache_key = f"mode_{config.module_name}"
        if cache_key in ai_config._cache:
            del ai_config._cache[cache_key]
        
        # 获取实际运行模式
        actual_mode = ai_config.get_module_mode(config.module_name)
        
        logger.info(f"模块 {config.module_name} 配置已更新: enabled={config.enabled}, mode={config.mode}, actual={actual_mode}")
        
        return {
            'success': True,
            'message': f'模块 {config.module_name} 配置已更新',
            'data': {
                'module_name': config.module_name,
                'enabled': config.enabled,
                'configured_mode': config.mode,
                'actual_mode': actual_mode
            }
        }
        
    except Exception as e:
        logger.error(f"更新模块配置失败: {e}")
        raise HTTPException(status_code=500, detail=f"更新模块配置失败: {str(e)}")

@router.post(ROUTES.ai.cache_clear)
async def clear_model_cache():
    """清理模型缓存"""
    try:
        model_cache = get_model_cache_manager()
        
        # 获取清理前的缓存信息
        before_info = model_cache.get_cache_info()
        
        # 清理缓存
        success = model_cache.clear_cache()
        
        if success:
            # 获取清理后的缓存信息
            after_info = model_cache.get_cache_info()
            
            return {
                'success': True,
                'message': '模型缓存已清理',
                'data': {
                    'before': before_info,
                    'after': after_info,
                    'freed_mb': before_info.get('cache_size_mb', 0)
                }
            }
        else:
            raise HTTPException(status_code=500, detail="缓存清理失败")
            
    except Exception as e:
        logger.error(f"清理模型缓存失败: {e}")
        raise HTTPException(status_code=500, detail=f"清理模型缓存失败: {str(e)}")

@router.post(ROUTES.ai.lightweight_train)
async def train_lightweight_model():
    """训练轻量级模型"""
    try:
        lightweight_filter = get_lightweight_filter()
        
        # 这里可以加载训练数据
        # 示例：从训练数据文件加载
        ad_samples = []
        normal_samples = []
        
        # 可以从数据库或文件加载训练样本
        # ad_samples = load_ad_samples()
        # normal_samples = load_normal_samples()
        
        if not ad_samples and not normal_samples:
            return {
                'success': False,
                'message': '缺少训练数据，请先收集足够的广告和正常内容样本'
            }
        
        # 训练模型
        success = lightweight_filter.train_with_samples(ad_samples, normal_samples)
        
        if success:
            return {
                'success': True,
                'message': '轻量级模型训练完成',
                'data': {
                    'ad_samples_count': len(ad_samples),
                    'normal_samples_count': len(normal_samples)
                }
            }
        else:
            return {
                'success': False,
                'message': '轻量级模型训练失败'
            }
            
    except Exception as e:
        logger.error(f"训练轻量级模型失败: {e}")
        raise HTTPException(status_code=500, detail=f"训练轻量级模型失败: {str(e)}")

@router.get(ROUTES.ai.recommendations)
async def get_ai_recommendations():
    """获取AI配置建议"""
    try:
        ai_config = get_ai_config()
        
        dependencies = {
            'sentence_transformers': False,  # 已移除
        }
        
        recommendations = _get_recommendations(ai_config, dependencies)
        
        return {
            'success': True,
            'data': {
                'recommendations': recommendations
            }
        }
        
    except Exception as e:
        logger.error(f"获取AI建议失败: {e}")
        raise HTTPException(status_code=500, detail=f"获取AI建议失败: {str(e)}")

def _get_recommendations(ai_config, dependencies: Dict[str, bool]) -> List[Dict[str, Any]]:
    """生成配置建议"""
    recommendations = []
    
    # 检查sentence_transformers可用性
    if not dependencies.get('sentence_transformers', False):
        recommendations.append({
            'type': 'warning',
            'title': '深度学习模式已废弃',
            'message': '系统已优化为轻量级模式，无需重量级依赖',
            'action': {
                'type': 'install',
                'command': 'pip install sentence-transformers',
                'description': '安装深度学习依赖'
            }
        })
    
    # 检查是否有模块使用了不可用的深度模式
    for module_name, config in ai_config.ai_modules.items():
        if config['mode'] == 'deep' and not dependencies.get('sentence_transformers', False):
            recommendations.append({
                'type': 'error',
                'title': f'模块{module_name}配置问题',
                'message': f'模块设置为deep模式但系统已改为轻量级',
                'action': {
                    'type': 'config',
                    'module': module_name,
                    'suggested_mode': 'lightweight',
                    'description': '切换到轻量级模式'
                }
            })
    
    # 性能建议
    if dependencies.get('sentence_transformers', False):
        recommendations.append({
            'type': 'info',
            'title': '性能优化建议',
            'message': '系统已优化，推荐使用auto或lightweight模式',
            'action': {
                'type': 'config',
                'suggested_mode': 'auto',
                'description': '使用自动模式'
            }
        })
    else:
        recommendations.append({
            'type': 'info',
            'title': '轻量级模式可用',
            'message': '轻量级模式提供良好的性能和准确率，无需额外依赖',
            'action': {
                'type': 'config',
                'suggested_mode': 'lightweight',
                'description': '使用轻量级模式'
            }
        })
    
    return recommendations